File: poh1/poh1_2.java
import java.util.*;
class Main{
	static final int SIZE=9999999;
	static byte[] z=new byte[SIZE];
	static int[] _v=new int[1000001],v;
	static int input_count=0;

	static int get(){
		int r=0;
		for(;'0'<=z[input_count]&&z[input_count]<='9';)r=r*10+z[input_count++]-'0';
		input_count++;
		return r;
	}
	public static void main(String[]Z){try{
		System.in.read(z,0,SIZE);
		int n,d,m,i,j,k,r;
		n=get();d=get();
		v=new int[n];
		for(i=0;i<n;i++)_v[get()]++;
		for(i=j=0;j<1000001;j++)for(k=0;k<_v[j];k++)v[i++]=j;
		for(i=0;i<d;i++){
			m=get();
			int idx=Arrays.binarySearch(v,m-v[0]+1);
			if(idx<0)idx=~idx;
			for(r=j=0,k=idx-1;r<m&&j<k&&v[j]+v[j+1]<=m;j++){
				for(;v[j]+v[k]>m;)k--;
				if(r<v[j]+v[k])r=v[j]+v[k];
			}
			System.out.println(r);
		}
	}catch(java.io.IOException e){}}
}

File: poh1/poh1_2.kt
import java.util.*
val SIZE=9999999
val z=ByteArray(SIZE)
val _v=IntArray(1000001)
var input_count=0

fun get():Int{
	var r:Int=0
	while('0'.toByte()<=z[input_count]&&z[input_count]<='9'.toByte())r=r*10+z[input_count++]-'0'.toByte()
	input_count++
	return r
}
fun main(args: Array<String>){try{
	System.`in`.read(z,0,SIZE)
	var n=get()
	var d=get()
	var m:Int
	var r:Int
	var v=IntArray(n)
	for(i in 0..n-1)_v[get()]++
	var i:Int=0
	for(j in 0..1000000){
		for(k in 0.._v[j]-1)v[i++]=j
	}
	for(i in 0..d-1){
		m=get()
		var idx:Int=Arrays.binarySearch(v,m-v[0]+1)
		if(idx<0){idx=-1 xor idx}
		r=0
		var j:Int=0
		var k:Int=idx-1
		while(r<m&&j<k&&v[j]+v[j+1]<=m){
			while(v[j]+v[k]>m)k--;
			if(r<v[j]+v[k])r=v[j]+v[k];
			j++
		}
		println(r);
	}
}catch(e:java.io.IOException){}}
